src/Discord.Net.Interactions/Info/Commands/CommandInfo.cs

```diff
@@ -0,0 +1,249 @@
+        protected abstract Task InvokeModuleEvent (IInteractionCommandContext context, IResult result);
+        protected abstract string GetLogString (IInteractionCommandContext context);
+
+        /// <inheritdoc/>
+        public async Task<PreconditionResult> CheckPreconditionsAsync (IInteractionCommandContext context, IServiceProvider services)
+        {
+            async Task<PreconditionResult> CheckGroups (ILookup<string, PreconditionAttribute> preconditions, string type)
+            {
+                foreach (IGrouping<string, PreconditionAttribute> preconditionGroup in preconditions)
+                {
+                    if (preconditionGroup.Key == null)
+                    {
+                        foreach (PreconditionAttribute precondition in preconditionGroup)
+                        {
+                            var result = await precondition.CheckRequirementsAsync(context, this, services).ConfigureAwait(false);
+                            if (!result.IsSuccess)
+                                return result;
+                        }
+                    }
+                    else
+                    {
+                        var results = new List<PreconditionResult>();
+                        foreach (PreconditionAttribute precondition in preconditionGroup)
+                            results.Add(await precondition.CheckRequirementsAsync(context, this, services).ConfigureAwait(false));
+
+                        if (!results.Any(p => p.IsSuccess))
+                            return PreconditionGroupResult.FromError($"{type} precondition group {preconditionGroup.Key} failed.", results);
+                    }
+                }
+                return PreconditionGroupResult.FromSuccess();
+            }
+
+            var moduleResult = await CheckGroups(Module.GroupedPreconditions, "Module").ConfigureAwait(false);
+            if (!moduleResult.IsSuccess)
+                return moduleResult;
+
+            var commandResult = await CheckGroups(_groupedPreconditions, "Command").ConfigureAwait(false);
+            if (!commandResult.IsSuccess)
+                return commandResult;
+
+            return PreconditionResult.FromSuccess();
+        }
+
+        protected async Task<IResult> RunAsync (IInteractionCommandContext context, object[] args, IServiceProvider services)
+        {
+            services ??= EmptyServiceProvider.Instance;
+
+            switch (RunMode)
+            {
+                case RunMode.Sync:
+                    {
+                        using var scope = services.CreateScope();
+                        return await ExecuteInternalAsync(context, args, scope.ServiceProvider).ConfigureAwait(false);
+                    }
+                case RunMode.Async:
+                    _ = Task.Run(async ( ) =>
+                    {
+                        using var scope = services.CreateScope();
```

Comment:
Throws InvalidOperationException

```
No service for type 'Microsoft.Extensions.DependencyInjection.IServiceScopeFactory' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.CreateScope(IServiceProvider provider)
   at Discord.Interactions.CommandInfo`1.<>c__DisplayClass37_0.<<RunAsync>b__0>d.MoveNext() in C:\Users\lynch\Documents\GitHub\Discord.Net Labs\Discord.Net-Labs\src\Discord.Net.Interactions\Info\Commands\CommandInfo.cs:line 142

Reply:
I don't think this is even necessary to create a new scoped service provider.

Reply:
Im guessing you are trying to use the service without creating a `ServiceProvider`, cuz normally `IServiceScopeFactory` should be resolveable without explicitly injecting it. I will try to fix this by adding a psuedo `IServiceScopeFactory` to the EmptyServiceProvider class.

Reply:
Following the conventions of Asp.net controllers, i think it is really useful to be able to create scoped service instances per request. Also there is no downside to providing the users with the option to use scoped services. If you are just using singletons, you wont be affected by this.